uri/iri.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	if u.Scheme != "xmpp" {
+		return nil, errBadScheme
+	}
+
+	uri := &URI{
+		URL: u,
+	}
+
+	if u.Host != "" {
+		// If an authentication address was provided (ie. the URI started with
+		// `xmpp://'), parse it out and take the recipient address from the path.
+
+		uri.AuthAddr, err = jid.New(u.User.Username(), u.Hostname(), "")
+		if err != nil {
+			return nil, err
+		}
+		if u.Path != "" {
+			// Strip the root / and use the path as the JID.
+			iri, err := toIRI(u.Path[1:], false)
+			if err != nil {
+				return nil, err
+			}
+			uri.ToAddr, err = jid.Parse(iri)
+			if err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		// If no auth address was provided (ie. the URI started with `xmpp:') take
+		// the recipient address from the opaque part and ignore the user info.
+		iri, err := toIRI(u.Opaque, true)
+		if err != nil {
+			return nil, err
+		}
+		uri.ToAddr, err = jid.Parse(iri)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for k, v := range u.Query() {
+		if len(v) == 0 || len(v) == 1 && v[0] == "" {
+			uri.Action = k
+			break
+		}
+	}
+
+	return uri, err
+}
+
+// String reassembles the URI or IRI Into a valid IRI string.
+func (u *URI) String() string {
+	iri, _ := toIRI(u.URL.String(), true)
+	return iri
+}
+
+// toIRI converts the URI to a valid IRI using the algorithm defined in RFC 3987
+// §3.2.
+// It does not validate that the input is a valid URI.
+func toIRI(u string, needsUnescape bool) (string, error) {
+	// 1.  Represent the URI as a sequence of octets in US-ASCII.
+	//
+	// 2.  Convert all percent-encodings ("%" followed by two hexadecimal
+	//     digits) to the corresponding octets, except those corresponding
+	//     to "%", characters in "reserved", and characters in US-ASCII not
+	//     allowed in URIs.
+	// TODO: using PathUnescape to create a new string is very inefficient, but
+	// it's the only method available in the standard library for this.
+	// In the future we should write an escape/unescaper that implements
+	// "golang.org/x/text/transform".Transformer or simply appends to a buffer or
+	// byte slice so that the next step can also be done in the same iteration
+	// without creating yet another builder.
+	var err error
+	if needsUnescape {
+		u, err = url.PathUnescape(u)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// 3. Re-percent-encode any octet produced in step 2 that is not part
+	//    of a strictly legal UTF-8 octet sequence.
+	// 4. Re-percent-encode all octets produced in step 3 that in UTF-8
+	//    represent characters that are not appropriate according to
+	//    sections 2.2, 4.1, and 6.1.
+	u = escapeInvalidUTF8(u)
+
+	// 5. Interpret the resulting octet sequence as a sequence of characters
+	//    encoded in UTF-8.
+	// TODO:
+
+	return u, nil
+}
+
+// escapeInvalidUTF8 is like strings.ToValidUTF8 except that it replaces invalid
+// UTF8 with % encoded versions of the invalid bytes instead of a fixed string.
+func escapeInvalidUTF8(s string) string {
```

Comment:
This needs to be rewritten before merge; this got copied, then its purpose changed and a bunch of stuff got tacked on, and now it's a monstrosity.

Reply:
Wrote a nicer implementation that's much easier to read, but it's far less optimized than the version from the standard library (as you might expect) and I can't see a good way to get similar performance (mine was an ~3.5x slowdown) and the same readability. Given that it wasn't even all that much easier to follow, I'll stick with this implementation for now while I think about it more.